### PR TITLE
SMS (SNS) stored queries from CWI into Terraform / SMS (pinpoint) new queries

### PR DESCRIPTION
**Pull request by jimleroyer**
# Summary | Résumé

* Pushes existing SMS / SNS variant queries from CloudWatch Logs Insight into Terraform IaC.
* Creates new SMS / Pinpoint variant queries in both CloudWatch Logs Insight and Terraform IaC.

## Related Issues | Cartes liées

* [Cloudwatch query / alert for success/failure rate by providers](https://app.zenhub.com/workspaces/notify-planning-core-6411dfb7c95fb80014e0cab0/issues/gh/cds-snc/notification-planning-core/431)

## Before merging this PR

Read code suggestions left by the
[cds-ai-codereviewer](https://github.com/cds-snc/cds-ai-codereviewer/) bot. Address
valid suggestions and shortly write down reasons to not address others. To help
with the classification of the comments, please use these reactions on each of the
comments made by the AI review:

| Classification      | Reaction | Emoticon |
|---------------------|----------|----------|
| Useful              | +1       | 👍        |
| Noisy               | eyes     | 👀        |
| Hallucination       | confused | 😕        |
| Wrong but teachable | rocket   | 🚀        |
| Wrong and incorrect | -1       | 👎        |

The classifications will be extracted and summarized into an analysis of how helpful
or not the AI code review really is.

## Test instructions | Instructions pour tester la modification

Before the PR merge, make sure the plan passes. After the merge, check the CloudWatch Logs Insight queries are created and corresponding alarms.

## Release Instructions | Instructions pour le déploiement

None.

## Reviewer checklist | Liste de vérification du réviseur

* [ ] This PR does not break existing functionality.
* [ ] This PR does not violate GCNotify's privacy policies.
* [ ] This PR does not raise new security concerns. Refer to our GC Notify Risk Register document on our Google drive.
* [ ] This PR does not significantly alter performance.
* [ ] Additional required documentation resulting of these changes is covered (such as the README, setup instructions, a related ADR or the technical documentation).

> ⚠ If boxes cannot be checked off before merging the PR, they should be moved to the "Release Instructions" section with appropriate steps required to verify before release. For example, changes to celery code may require tests on staging to verify that performance has not been affected.

### CODE REVIEW

File: aws/pinpoint_to_sqs_sms_callbacks/cloudwatch_queries.tf [hunk 1]
```diff
@@ -1,6 +1,41 @@
...
+  query_string = <<QUERY
+stats avg((eventTimestamp - messageRequestTimestamp) / 1000 / 60) as Avg_carrier_time_minutes,
+count(*) as Number by carrierName as Carrier
+| filter isFinal
+| sort by Avg_carrier_time_minutes desc
+| limit 100
+QUERY
+}
+
+resource "aws_cloudwatch_query_definition" "pinpoint-failures-by-carrier" {
```
Comment (jimleroyer) [inline aws/pinpoint_to_sqs_sms_callbacks/cloudwatch_queries.tf:19]: This is the new query over here for failures by carrier.

File: aws/pinpoint_to_sqs_sms_callbacks/cloudwatch_queries.tf [hunk 2]
```diff
@@ -1,6 +1,41 @@
+resource "aws_cloudwatch_query_definition" "pinpoint-carrier-dwell-times" {
```
Comment (jimleroyer) [inline aws/pinpoint_to_sqs_sms_callbacks/cloudwatch_queries.tf:1]: This is the equivalent of `sms-carrier-dwell-times` but for pinpoint.

File: aws/common/cloudwatch_queries.tf [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+  log_group_names = [
+    aws_cloudwatch_log_group.sns_deliveries[0].name,
+    aws_cloudwatch_log_group.sns_deliveries_failures[0].name
+  ]
+
+  query_string = <<QUERY
+fields @timestamp as Timestamp, status as Status, notification.messageId as `Message ID`,
+    delivery.destination as `Destination phone number`, delivery.providerResponse as `Provider response`,
+    delivery.smsType as `Message type`
+| filter delivery.destination like '1416xxxxxxx'
```
[the remark this answers is not in the document]
Comment (jimleroyer) [reply aws/common/cloudwatch_queries.tf:62]: I don't think we can use parameters in these unfortunately. If that was a query in a dashboard, yeah but that is not the case.

File: aws/pinpoint_to_sqs_sms_callbacks/cloudwatch_alarms.tf [hunk 4]
```diff
@@ -149,6 +149,195 @@ resource "aws_cloudwatch_metric_alarm" "total-sms-spending-critical" {
   }
 }
 
+resource "aws_cloudwatch_metric_alarm" "pinpoint-sms-failures-bell-warning" {
+  count               = var.cloudwatch_enabled ? 1 : 0
+  alarm_name          = "pinpoint-sms-failures-telus-warning"
```
[the remark this answers is not in the document]
Comment (jimleroyer) [reply aws/pinpoint_to_sqs_sms_callbacks/cloudwatch_alarms.tf:154]: I really need the AI with my clumsiness.

File: aws/pinpoint_to_sqs_sms_callbacks/cloudwatch_logs.tf [hunk 5]
```diff
@@ -3,6 +3,7 @@
 #
 
 resource "aws_cloudwatch_log_group" "pinpoint_deliveries" {
+  # REVIEW: We might want the count attribute present to disable this resource
```
Comment (jimleroyer) [inline aws/pinpoint_to_sqs_sms_callbacks/cloudwatch_logs.tf:6]: @ben851 @sastels Was it a mistake to not add a switch to enable/disable the resource? The equivalent sns log group has the switch, so I assume it should be over here too.

File: aws/pinpoint_to_sqs_sms_callbacks/cloudwatch_logs.tf [hunk 6]
```diff
@@ -11,6 +12,7 @@ resource "aws_cloudwatch_log_group" "pinpoint_deliveries" {
 }
 
 resource "aws_cloudwatch_log_group" "pinpoint_deliveries_failures" {
+  # REVIEW: We might want the count attribute present to disable this resource
```
Comment (jimleroyer) [inline aws/pinpoint_to_sqs_sms_callbacks/cloudwatch_logs.tf:15]: @ben851 @sastels Was it a mistake to not add a switch to enable/disable the resource? The equivalent sns log group has the switch, so I assume it should be over here too.

File: aws/common/cloudwatch_queries.tf [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+
+  log_group_names = [
+    aws_cloudwatch_log_group.sns_deliveries[0].name,
+    aws_cloudwatch_log_group.sns_deliveries_failures[0].name
+  ]
+
+  query_string = <<QUERY
+stats avg(delivery.dwellTimeMsUntilDeviceAck / 1000 / 60) as Avg_carrier_time_minutes, 
+| count(*) as Number by delivery.phoneCarrier as Carrier
+QUERY
```
[the remark this answers is not in the document]
Comment (jimleroyer) [reply aws/common/cloudwatch_queries.tf:30]: It is on a new line.

File: aws/pinpoint_to_sqs_sms_callbacks/cloudwatch_logs.tf (hunk 5 above)
Comment (ben851) [reply aws/pinpoint_to_sqs_sms_callbacks/cloudwatch_logs.tf:6]: Yeah probably a good idea to have it enabled/disabled based on cloudwatch enabled.

File: aws/pinpoint_to_sqs_sms_callbacks/cloudwatch_logs.tf (hunk 6 above)
Comment (ben851) [reply aws/pinpoint_to_sqs_sms_callbacks/cloudwatch_logs.tf:15]: Ditto

**Review by ben851 — APPROVED** [review]
LGTM but feel free to add the count line on those two log groups

**Pull request merged**